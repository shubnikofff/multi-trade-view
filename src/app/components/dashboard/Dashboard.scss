@import "../../variables";

.dashboard {
  display: flex;
  flex-direction: column;
  align-items: center;

  & > * {
    padding: $gutter;
  }
}
