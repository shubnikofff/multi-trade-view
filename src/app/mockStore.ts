import configureMockStore from 'redux-mock-store'
import { UserRole } from './types/user';

const mockStore = configureMockStore([]);

const store = mockStore({
    auth: UserRole.Seller,
    trades: {
        ids: [1, 2],
        entities: {
            1: {
                id: 1,
                chatId: 1,
                paymentMethod: 'Amazon Gift Card',
                paid: false,
                buyerId: 1,
                amount: 100,
                hash: 'abc42',
                started: new Date(2020, 8),
            },
            2: {
                id: 2,
                chatId: 2,
                paymentMethod: 'iTunes Gift Card',
                paid: true,
                buyerId: 1,
                amount: 150,
                hash: 'cd23e',
                started: new Date(2020, 8),
            }
        }
    },
    users: {
        ids: [1],
        entities: {
            1: {
                id: 1,
                avatarUrl: 'url',
                name: 'username',
                positiveReputation: 25,
                negativeReputation: 12,
                ofTrades: 4,
            }
        }
    },
    chats: {
        ids: [1, 2],
        entities: {
            1: {
                id: 1,
                hasUnreadMessages: false,
                messages: [
                    {
                        message: 'Message_1',
                        sender: UserRole.Buyer,
                        sendTime: new Date(2020, 8),
                    },
                    {
                        message: 'Message_2',
                        sender: UserRole.Seller,
                        sendTime: new Date(2020, 8),
                    }
                ],
            },
            2: {
                id: 2,
                hasUnreadMessages: false,
                messages: [
                    {
                        message: 'Message_1',
                        sender: UserRole.Buyer,
                        sendTime: new Date(2020, 8),
                    },
                    {
                        message: 'Message_2',
                        sender: UserRole.Seller,
                        sendTime: new Date(2020, 8),
                    }
                ],
            },
        },
    },
    rate: 10000,
});

export { store }
