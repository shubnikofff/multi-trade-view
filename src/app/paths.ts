export const DASHBOARD = '/dashboard/:tradeId';
export const CHAT = '/chat/:tradeId';
export const ROOT = '/';
export const TRADE = '/trade/:tradeId';
