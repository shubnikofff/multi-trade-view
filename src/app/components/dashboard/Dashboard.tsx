import React from 'react';
import classNames from 'classnames';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons'
import { Link } from 'react-router-dom';
import Avatar from '@components/avatar';
import NotAvailable from '@components/not-available';

import { UserRoleEnum } from '@type/User';

import { PATH_CHAT } from '@app/paths';

import { useDashboard } from './useDashboard';

import './Dashboard.scss';

interface DashboardProps {
    smallScreen?: boolean;
}

function Dashboard({ smallScreen }: DashboardProps) {
    const {
        auth,
        buyer,
        convert,
        trade,
        releaseBitcoins,
    } = useDashboard();

    if (auth === UserRoleEnum.Buyer) {
        return (
            <NotAvailable>
                Not available for this user
            </NotAvailable>
        )
    }

    if (!trade) {
        return (
            <NotAvailable>
                No such trade
            </NotAvailable>
        );
    }

    return (
        <div className="dashboard">
            <div className="dashboard__top">
                {smallScreen && <div className="dashboard__navigation-area">
                    <Link to={`${PATH_CHAT}/${trade.hash}`}>
                        <FontAwesomeIcon icon={faChevronLeft} color="black"/>
                    </Link>
                </div>}
                <div>
                    You are trading with <b>{buyer?.name}</b>
                </div>
                <div className="dashboard__start-time">
                    Started {new Date(trade.started).toLocaleString()}
                </div>
                <div className="dashboard__action-area">
                    <button
                        className="btn"
                        disabled={trade.paid}
                        onClick={releaseBitcoins}
                    >
                        Release bitcoins
                    </button>
                </div>
            </div>
            <table className="dashboard__bottom">
                <tbody>
                <tr>
                    <td>
                        <Avatar url={buyer?.avatarUrl}/>
                        <div className="chat__header__trade-info__buyer">
                            <span className="text-primary">+{buyer?.positiveReputation}</span>
                            <span> / </span>
                            <span className="text-danger">-{buyer?.negativeReputation}</span>
                        </div>
                    </td>
                    <td>
                        <div className="dashboard__bottom_label">
                            <b># Of trades</b>
                        </div>
                        <div>4</div>
                    </td>
                </tr>
                <tr>
                    <td>
                        <div className="dashboard__bottom_label">
                            <b>Trade status</b>
                        </div>
                        <div className={classNames('dashboard__status', trade.paid
                            ? 'dashboard__status_green'
                            : 'dashboard__status_gray'
                        )}>
                            {trade.paid ? 'Paid' : 'Not paid'}
                        </div>
                    </td>
                    <td>
                        <div className="dashboard__bottom_label">
                            <b>Trade hash</b>
                        </div>
                        <div className="text-secondary">
                            {trade.hash}
                        </div>
                    </td>
                </tr>
                <tr>
                    <td>
                        <div className="dashboard__bottom_label">
                            <b>Amount USD</b>
                        </div>
                        <div>{trade.amount.toFixed(2)}</div>
                    </td>
                    <td>
                        <div className="dashboard__bottom_label">
                            <b>Amount BTC</b>
                        </div>
                        <div className="text-secondary">
                            {convert(trade.amount).toFixed(8)}
                        </div>
                    </td>
                </tr>
                </tbody>
            </table>
        </div>
    );
}

export default React.memo<DashboardProps>(Dashboard);
