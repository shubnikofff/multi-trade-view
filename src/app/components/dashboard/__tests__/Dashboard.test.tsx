import React from 'react';

import { mount } from 'enzyme';
import { createMemoryHistory } from 'history'

import { Provider } from 'react-redux';
import { Router, Route } from 'react-router-dom'
import Dashboard from '../Dashboard';

import { store } from '../../../mockStore';
import { PATH_DASHBOARD } from '../../../paths';

describe('Component Dashboard', () => {
    const history = createMemoryHistory({
        initialEntries: [`${PATH_DASHBOARD}/1`],
        keyLength: 0,
    });

    const wrapper = mount(
        <Provider store={store}>
            <Router history={history}>
                <Route path={`${PATH_DASHBOARD}/:tradeId`}>
                    <Dashboard />
                </Route>
            </Router>
        </Provider>
    );

    afterEach(() => {
       store.clearActions();
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should trigger action updateTrade with paid true payload', () => {
        const expectedAction = {
            payload: { changes: { paid: true }, id: '1' },
            type: "trades/updateTrade",
        }

        wrapper.find('button').simulate('click');
        expect(store.getActions()).toEqual(expect.arrayContaining([expectedAction]));
    });
});
