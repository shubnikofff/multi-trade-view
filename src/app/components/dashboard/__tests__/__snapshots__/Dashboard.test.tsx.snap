// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Dashboard should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/dashboard/1",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/dashboard/1",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route
      path="/dashboard/:tradeId"
    >
      <Memo(Dashboard)>
        <div
          className="dashboard"
        >
          <div
            className="dashboard__top"
          >
            <div>
              You are trading with 
              <b>
                username
              </b>
            </div>
            <div
              className="dashboard__start-time"
            >
              Started 
              9/1/2020, 00:00:00
            </div>
            <div
              className="dashboard__action-area"
            >
              <button
                className="btn"
                disabled={false}
                onClick={[Function]}
              >
                Release bitcoins
              </button>
            </div>
          </div>
          <table
            className="dashboard__bottom"
          >
            <tbody>
              <tr>
                <td>
                  <Avatar
                    url="url"
                  >
                    <img
                      alt="Avatar"
                      className="avatar"
                      src="url"
                    />
                  </Avatar>
                  <div
                    className="chat__header__trade-info__buyer"
                  >
                    <span
                      className="text-primary"
                    >
                      +
                      25
                    </span>
                    <span>
                       / 
                    </span>
                    <span
                      className="text-danger"
                    >
                      -
                      12
                    </span>
                  </div>
                </td>
                <td>
                  <div
                    className="dashboard__bottom_label"
                  >
                    <b>
                      # Of trades
                    </b>
                  </div>
                  <div>
                    4
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  <div
                    className="dashboard__bottom_label"
                  >
                    <b>
                      Trade status
                    </b>
                  </div>
                  <div
                    className="dashboard__status dashboard__status_gray"
                  >
                    Not paid
                  </div>
                </td>
                <td>
                  <div
                    className="dashboard__bottom_label"
                  >
                    <b>
                      Trade hash
                    </b>
                  </div>
                  <div
                    className="text-secondary"
                  >
                    abc42
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  <div
                    className="dashboard__bottom_label"
                  >
                    <b>
                      Amount USD
                    </b>
                  </div>
                  <div>
                    100.00
                  </div>
                </td>
                <td>
                  <div
                    className="dashboard__bottom_label"
                  >
                    <b>
                      Amount BTC
                    </b>
                  </div>
                  <div
                    className="text-secondary"
                  >
                    0.01000000
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </Memo(Dashboard)>
    </Route>
  </Router>
</Provider>
`;
